### DISCUSSION TITLE
[I] Split `broadcast` and `connector` plugins

### DISCUSSION THREAD

**Discussion by steffendietz** [roadrunner-server/roadrunner#704] [Ideas]
Would it make sense to extract the `broadcast` functionality from the `websockets` plugin?

For example when thinking about a (potential future) `sse` plugin, which is also connected to the same broadcast channels. Then this nesting of functionality could cause problems down the line.

What do you think? Am I overthinking this?

**Comment by rustatian** [comment]
Hey @steffendietz. I guess it would. But if you want to use single-node RR with `websockets`, you may use `websockets` with the `memory` plugin.
Or might be I have misunderstood your idea. Could you please describe your case in detail?

**Reply by steffendietz** [reply]
Currently `websockets` are used to let a browser connect to broadcast topics.
There is another method of *pushing* data to the browser called [server sent events](https://developer.mozilla.org/en-US/docs/Web/API/EventSource).
Similar to websockets the connection is not closed, but kept open, so that the server can continue to push data to the client.

I realize that roadrunner does not currently have SSE support, but I was just pondering if it would make sense to detach the internal *broadcast* handling (connecting to memory or redis broker) from the delivery method (websockets, sse, ...).

**Reply by rustatian** [reply]
Got u. You are lucky 😄 , because I have designed a broadcast (pub-sub) interface without bounding it to the only WebSocket protocol. You can have any `Subscriber` (`sse` in your case) with any desired storage (even just raw memory) under the hood. From my POV, this is a good idea to support SSE in the RR2. If you don't mind, could you please create a feature request with a link to this discussion and description?

**Reply by steffendietz** [reply]
Okay. Great. Yes. I will do so.

The `broadcast` plugin of RoadRunner `1.0` also implemented the *drivers* (memory, redis, ...) and the `broadcast-ws` plugin then internally connected to the `broadcast` plugin.

So I thought seperating this again, so that you can have a future `broadcast-sse` plugin would make sense.

I think I just didn't realize, that the `memory` and `redis` plugins now essentially *are* the broadcast plugin, since they implement the pub-sub interface.

Thanks for clearing this up. :)

**Comment by wolfy-j** [comment]
You should be able to use multiple brokers and pipelines to avoid interception. We plan to use broadcast for many other things within the RoadRunner.

**Comment by steffendietz** [comment]
Great news regarding #729. Best possible outcome I could hope for. 👍
